zkvm/src/types.rs

```diff
@@ -300,12 +342,61 @@ impl Value {
 }
 
 impl Expression {
-    pub fn constant<S: Into<Scalar>>(a: S) -> Self {
-        let a: Scalar = a.into();
+    pub fn constant<S: Into<ScalarWitness>>(a: S) -> Self {
+        Expression::Constant(a.into())
+    }
+}
+
+impl Neg for Expression {
+    type Output = Expression;
+
+    fn neg(self) -> Expression {
+        match self {
+            Expression::Constant(a) => Expression::Constant(-a),
+            Expression::Terms(mut terms, assignment) => {
+                for (_, n) in terms.iter_mut() {
+                    *n = -*n;
+                }
+
+                let x = match assignment {
+                    Some(a) => Some(-a),
+                    None => None,
+                };
+                Expression::Terms(terms, x)
+            }
+        }
+    }
+}
 
-        Expression {
-            terms: vec![(r1cs::Variable::One(), a)],
-            assignment: Some(ScalarWitness::Scalar(a)),
+impl Add for Expression {
+    type Output = Expression;
+
+    fn add(self, rhs: Expression) -> Expression {
+        match (self, rhs) {
+            (Expression::Constant(left), Expression::Constant(right)) => {
+                Expression::Constant(left + right)
+            }
+            (Expression::Constant(left), Expression::Terms(mut right_terms, right_assignment)) => {
+                // prepend constant term to `term vector` in non-constant expression
+                right_terms.insert(0, (r1cs::Variable::One(), left.into()));
+                Expression::Terms(right_terms, right_assignment.map(|b| left + b))
+            }
+            (Expression::Terms(mut left_terms, left_assignment), Expression::Constant(right)) => {
+                // append constant term to term vector in non-constant expression
+                left_terms.push((r1cs::Variable::One(), right.into()));
+                Expression::Terms(left_terms, left_assignment.map(|a| a + right))
+            }
+            (
+                Expression::Terms(mut left_terms, left_assignment),
+                Expression::Terms(right_terms, right_assignment),
+            ) => {
+                // append right terms to left terms in non-constant expression
+                left_terms.extend(right_terms);
+                Expression::Terms(
+                    left_terms,
+                    left_assignment.and_then(|a| right_assignment.map(|b| a + b)),
```

Comment:
```suggestion
                    left_assignment.and_then(|l| right_assignment.map(|r| l + r)),
```

sorry for a bikesheddy nitpick